relay-cardinality/src/limiter.rs

```diff
@@ -199,7 +200,7 @@ impl<T: Limiter> CardinalityLimiter<T> {
 #[derive(Debug, Default)]
 struct DefaultReporter<'a> {
     exceeded_limits: HashSet<&'a CardinalityLimit>,
-    entries: HashSet<usize>,
+    entries: HashMap<usize, SmallVec<[String; 4]>>,
```

Comment:
```suggestion
    entries: HashMap<usize, SmallVec<[&'a CardinalityLimit; 1]>>,
```

I think storing a reference to the limit should work here, like we do it for the `exceeded_limits`

Reply:
I would also prefer this, but it results in a _very annoying_ lifetime problem. To wit: in
```rust
    pub fn rejected(&self) -> impl Iterator<Item = &T> {
        self.rejections.keys().filter_map(|&i| self.source.get(i))
    }
```
the type of the unnamed iterator we return contains the type of the `rejections` map, which contains the lifetime `'a`. To which the compiler says
```
error[E0700]: hidden type for `impl Iterator<Item = &T>` captures lifetime that does not appear in bounds
   --> relay-cardinality/src/limiter.rs:300:9
    |
263 | impl<'a, T> CardinalityLimits<'a, T> {
    |      -- hidden type `FilterMap<hashbrown::hash_map::Keys<'_, usize, SmallVec<[&'a config::CardinalityLimit; 1]>>, {closure@relay-cardinality/src/limiter.rs:300:43: 300:47}>` captures the lifetime `'a` as defined here
...
299 |     pub fn rejected(&self) -> impl Iterator<Item = &T> {
    |                               ------------------------ opaque type defined here
300 |         self.rejections.keys().filter_map(|&i| self.source.get(i))
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
This is frustrating because the the `'a` really has nothing to do with what we're trying to return—it's in the value type of the map, but we only want to iterate over the keys.